src/main/java/io/vertx/httpproxy/impl/CachingFilter.java

```diff
@@ -102,49 +95,54 @@ private Future<ProxyResponse> tryHandleProxyRequestFromCache(ProxyContext contex
 
     HttpServerRequest response = proxyRequest.proxiedRequest();
 
-    Resource resource;
     HttpMethod method = response.method();
-    if (method == HttpMethod.GET || method == HttpMethod.HEAD) {
-      String cacheKey = proxyRequest.absoluteURI();
-      resource = cache.computeIfPresent(cacheKey, CACHE_GET_AND_VALIDATE);
+    if (method != HttpMethod.GET && method != HttpMethod.HEAD) {
+      return context.sendRequest();
+    }
+
+    String cacheKey = proxyRequest.absoluteURI();
+    return cache.get(cacheKey).compose(resource -> {
       if (resource == null) {
-        return null;
+        return context.sendRequest();
+      }
+
+      long now = System.currentTimeMillis();
+      long val = resource.getTimestamp() + resource.getMaxAge();
+      if (val < now) {
+        return cache.remove(cacheKey).compose(v -> context.sendRequest());
       }
-    } else {
-      return null;
-    }
 
-    String cacheControlHeader = response.getHeader(HttpHeaders.CACHE_CONTROL);
-    if (cacheControlHeader != null) {
-      CacheControl cacheControl = new CacheControl().parse(cacheControlHeader);
-      if (cacheControl.maxAge() >= 0) {
-        long now = System.currentTimeMillis();
-        long currentAge = now - resource.timestamp;
-        if (currentAge > cacheControl.maxAge() * 1000) {
-          String etag = resource.headers.get(HttpHeaders.ETAG);
-          if (etag != null) {
-            proxyRequest.headers().set(HttpHeaders.IF_NONE_MATCH, resource.etag);
-            context.set("cached_resource", resource);
-            return context.sendRequest();
-          } else {
-            return null;
```

Comment:
Was this a bug?

Reply:
Do you mean my version? Considering the method `handleProxyRequest`, every `return null` is identical to `return context.sendRequest()`. Possibly it's better to combine these two returns